algorithms/linfa-clustering/src/k_means/init.rs

```diff
@@ -0,0 +1,389 @@
+/// centroids.
+fn k_means_para<'a, F: Float + SampleUniform + for<'b> AddAssign<&'b F>>(
+    n_clusters: usize,
+    observations: &'a ArrayView2<'a, F>,
+    rng: &mut impl Rng,
+    rng_func: &RngFunc<F>,
+) -> Array2<F> {
+    // The product of these parameters must exceed n_clusters. The higher they are, the more
+    // candidates are selected, which improves the quality of the centroids but increases running
+    // time. The values provided here are "sweetspots" suggested by the paper.
+    let n_rounds = 8;
+    let candidates_per_round = n_clusters;
+
+    let (n_samples, n_features) = observations.dim();
+    let mut candidates = Array2::zeros((n_clusters * n_rounds, n_features));
+
+    // Pick 1st centroid randomly
+    let first_idx = rng.gen_range(0, n_samples);
+    candidates.row_mut(0).assign(&observations.row(first_idx));
+    let mut n_candidates = 1;
+
+    let mut dists = Array1::zeros(n_samples);
+    'outer: for _ in 0..n_rounds {
+        let current_candidates = candidates.slice(s![0..n_candidates, ..]);
+        update_min_dists(&current_candidates, &observations, &mut dists);
+        // Generate the next set of candidates from the input points, using the same probability
+        // formula as KMeans++. On average this generates candidates equal to
+        // `candidates_per_round`.
+        let next_candidates = sample_subsequent_candidates(
+            observations,
+            &dists,
+            F::from(candidates_per_round).unwrap(),
+            rng.gen_range(0, 100000),
+            rng_func,
+        );
+
+        // Append the newly generated candidates to the current cadidates, breaking out of the loop
+        // if too many candidates have been found
+        for candidate in next_candidates.into_iter() {
+            candidates.row_mut(n_candidates).assign(&candidate);
+            n_candidates += 1;
+            if n_candidates >= candidates.nrows() {
+                break 'outer;
+            }
+        }
+    }
+
+    let final_candidates = candidates.slice(s![0..n_candidates, ..]);
+    // Weigh the candidate centroids by the sizes of the clusters they form in the input points.
+    let weights = cluster_membership_counts(&final_candidates, &observations);
+
+    // The number of candidates is almost certainly higher than the number of centroids, so we
+    // recluster the candidates into the right number of centroids using weighted KMeans++.
+    weighted_k_means_plusplus(n_clusters, &final_candidates, &weights.view(), rng)
+}
+
+/// Generate candidate centroids by sampling each observation in parallel using a supplied random
+/// number generation function.
+fn sample_subsequent_candidates<'a, F: Float>(
+    observations: &'a ArrayView2<'a, F>,
+    dists: &Array1<F>,
+    multiplier: F,
+    seed: u64,
+    // Must return a number between 0 and 1.
+    // Easiest way to generate a random number in a Rayon thread, since cloning the actual RNG is
```

Comment:
IsaacRng is 2Kb so it's definitely not cheap to clone. Other pseudo RNGs like xoshiro (256 bytes) are pretty small and can be cloned. One of the envisioned use cases for `rng_func` is `ThreadRng`, which is not Sync (so it can't be cloned in different threads) but is cheap to instantiate in different threads. Also, a possible issue with cloning a seeded RNG across different threads instead of using `rng_func` is that it produces the same value for all threads unless the clones are reseeded, so we'd also need an API for seeding the cloned RNGs as well.

Reply:
I know that `rand::rngs::SmallRng` uses the xoshiro implementation underneath and should be small enough to clone by the number of CPUs. But the issue with re-seeding exists, there exists [ReesedingRng](https://docs.rs/rand/0.8.3/rand/rngs/adapter/struct.ReseedingRng.html) wrapper, but I don't understand why it requires a block based rng .. how would you avoid constant cloning with a closure?

Reply:
With the `rng_func` closure any cloning/instantiation is done in a spawned thread, so the sequential runtime should be fine assuming you have enough threads. `ReseedingRng` requires the wrapped rng to be seedable which is pretty restrictive.

Reply:
this is still more expansive than just creating a single rng for each thread a-prior. You should be able to seed multiple rngs from your main one with https://docs.rs/rand/0.8.3/rand/trait.SeedableRng.html#method.from_rng by restricting with e.g. `R::from_rng(&mut rng) where R: Rng + SeedableRng`

Reply:
What I meant was that `rng_func` creates its RNGs inside the spawned threads rather than the main thread, so the creation of N generators is parallelized (assuming you have enough cores). If I instead seed a rng for each thread in the main thread then the creation of N generators is done entirely sequentially, so it would be slower.